File: python/processor.py
from config import Config
from create_table import NextVersionToProcess
from event_parser import INDEXER_NAME, parse
from aptos.indexer.v1 import raw_data_pb2_grpc

import grpc
from aptos.indexer.v1 import raw_data_pb2

from sqlalchemy import create_engine
from sqlalchemy.orm import Session

import argparse
import json


parser = argparse.ArgumentParser()
parser.add_argument("-c", "--config", help="Path to config file", required=True)
args = parser.parse_args()
config = Config.from_yaml_file(args.config)

metadata = (
    ("x-aptos-data-authorization", config.processor_api_key),
    ("x-aptos-request-name", INDEXER_NAME),
)
options = [("grpc.max_receive_message_length", -1)]
engine = create_engine(config.db_connection_uri)

starting_version = 0
if config.starting_version != None:
    # Start from config's starting version if set
    starting_version = config.starting_version
else:
    # Start from next version to process in db
    with Session(engine) as session, session.begin():
        next_version_to_process_from_db = session.get(
            NextVersionToProcess, INDEXER_NAME
        )
        if next_version_to_process_from_db != None:
            starting_version = next_version_to_process_from_db.next_version

print(
    json.dumps(
        {
            "message": "Connected to the indexer grpc",
            "starting_version": starting_version,
        }
    )
)
# Connect to grpc
with grpc.insecure_channel(config.processor_endpoint, options=options) as channel:
    stub = raw_data_pb2_grpc.RawDataStub(channel)
    current_transaction_version = starting_version

    for response in stub.GetTransactions(
        raw_data_pb2.GetTransactionsRequest(starting_version=starting_version),
        metadata=metadata,
    ):
        chain_id = response.chain_id

        if chain_id != config.chain_id:
            raise Exception(
                "Chain ID mismatch. Expected chain ID is: "
                + str(config.chain_id)
                + ", but received chain ID is: "
                + str(chain_id)
            )
        print(
            json.dumps(
                {
                    "message": "Response received",
                    "starting_version": response.transactions[0].version,
                }
            )
        )
        transactions_output = response
        for transaction in transactions_output.transactions:
            transaction_version = transaction.version
            if transaction_version != current_transaction_version:
                raise Exception(
                    "Transaction version mismatch. Expected transaction version is: "
                    + str(current_transaction_version)
                    + ", but received transaction version is: "
                    + str(transaction_version)
                )

            parsed_objs = parse(transaction)

            with Session(engine) as session, session.begin():
                # Insert Events into database
                if parsed_objs is not None:
                    session.add_all(parsed_objs)

                # Update latest processed version
                session.merge(
                    NextVersionToProcess(
                        indexer_name=INDEXER_NAME,
                        next_version=current_transaction_version + 1,
                    )
                )

            if (current_transaction_version % 1000) == 0:
                print(
                    json.dumps(
                        {
                            "message": "Successfully processed transaction",
                            "last_success_transaction_version": current_transaction_version,
                        }
                    )
                )

            current_transaction_version += 1

File: typescript/processor.ts
import { ArgumentParser } from 'argparse';
import { parse } from './grpc_parser';
import * as services from './aptos/indexer/v1/raw_data_grpc_pb';
import * as indexerRawDataMessages from './aptos/indexer/v1/raw_data_pb';
import { Config } from './config';
import { Timer } from 'timer-node';
import { exit } from 'process';
import { ChannelCredentials, CallCredentials, Metadata } from '@grpc/grpc-js';

// A hack to override the http2 settings
class CustomChannelCred extends ChannelCredentials {
  constructor(callCredentials?: CallCredentials) {
    super();
  }

  compose(callCredentials: CallCredentials): never {
    throw new Error('Cannot compose insecure credentials');
  }

  _getConnectionOptions(): any {
    return {
      settings: {
        // This will increase the http2 frame size. Default is 16384, which is too small.
        maxFrameSize: 4194304,
        // The initial window size is overridden. We need to patch the grpc-js library to allow this.
        initialWindowSize: 4194304,
        maxHeaderListSize: 8192,
        enablePush: false,
        maxConcurrentStreams: 0,
      },
    };
  }

  _isSecure(): boolean {
    return false;
  }

  _equals(other: ChannelCredentials): boolean {
    return other instanceof CustomChannelCred;
  }
}

// Parse the config file
const parser = new ArgumentParser({
  description: 'Indexer Stream Client arguments',
});
parser.add_argument('-c', '--config', {
  help: 'Path to config file',
  required: true,
});
parser.add_argument('-p', '--perf', {
  help: 'Show performance metrics. Takes one value: number of transactions to process e.g. -p 1000',
  required: false,
});
const args = parser.parse_args();
const config = Config.from_yaml_file(args.config);

// Create client and request
const client = new services.RawDataClient(
  config.indexer_endpoint,
  new CustomChannelCred(),
  {
    'grpc.keepalive_time_ms': 1000,
    'grpc.default_compression_algorithm': 2,
    'grpc.default_compression_level': 3,
    'grpc.max_receive_message_length': -1,
    'grpc.max_send_message_length': -1,
  }
);

const request = new indexerRawDataMessages.GetTransactionsRequest();
request.setStartingVersion(config.starting_version);
const metadata = new Metadata();
metadata.set('x-aptos-data-authorization', config.indexer_api_key);

// Create and start the streaming RPC
let currentTransactionVersion = config.starting_version;
const stream = client.getTransactions(request, metadata);

const timer = new Timer();
timer.start();

stream.on(
  'data',
  function (response: indexerRawDataMessages.TransactionsResponse) {
    const transactionsList = response.getTransactionsList();

    if (transactionsList == null) {
      return;
    }

    // Validate response chain ID matches expected chain ID
    if (response.getChainId() != config.chain_id) {
      throw new Error(
        `Chain ID mismatch. Expected ${
          config.chain_id
        } but got ${response.getChainId()}`
      );
    }

    if (transactionsList == null) {
      return;
    }

    for (const transaction of transactionsList) {
      // Validate transaction version is correct
      if (transaction.getVersion() != currentTransactionVersion) {
        throw new Error(
          `Transaction version mismatch. Expected ${currentTransactionVersion} but got ${transaction.getVersion()}`
        );
      }

      parse(transaction);

      if (currentTransactionVersion % 1000 == 0) {
        console.log({
          message: 'Successfully processed transaction',
          last_success_transaction_version: currentTransactionVersion,
        });
      }

      const totalTransactions =
        currentTransactionVersion - config.starting_version + 1;

      if (args.perf && totalTransactions >= args.perf) {
        timer.stop();
        console.log(
          `Takes ${timer.ms()} ms to receive ${totalTransactions} txns`
        );
        exit(0);
      }

      currentTransactionVersion += 1;
    }
  }
);

stream.on('error', function (e) {
  console.log(e);
  // An error has occurred and the stream has been closed.
});
stream.on('status', function (status) {
  console.log(status);
  // process status
});
